File: Data-Structures-And-Algorithms/Sort/BubbleSort.java
/**
 * 为什么插入排序要比冒泡排序效率高？
 * 插入排序相比冒泡排序元素交换的次数是一样的，不过每一次元素交换时，插入排序只需要一次复制操作，
 * 而冒泡排序需要三次赋值操作，赋值操作上花费时间是插入排序的三倍。
 */

// 插入排序：
public static void insertSort(int[] arr) {
    // 空数组 或 只有一个元素的数组，则什么都不做。
    if (arr == null || arr.length == 0) return;
    // 外层循环的 i 代表有序区的最后一个元素。
    for (int i = 0; i < arr.length - 1; i++) {
        // 待插入的元素暂存到value.
        int value = arr[i + 1];
        int j = i;
        // j < 0 时退出循环，说明 j 是最小的元素的索引值。
        // 或者 arr[j] <= value 时退出循环，说明 j 是比value小的元素的索引值。
        for (; j >= 0 && arr[j] > value; j--) {
            // 把元素往后挪。
            arr[j + 1] = arr[j];
        }
        // 把待插入元素，放到正确位置。
        arr[j + 1] = value;
    }
}

// 冒泡排序：
public static void BubbleSort(int[] arr) {
    int temp;
    for(int i=0;i<arr.length-1;i++){//冒泡趟数
        for(int j=0;j<arr.length-i-1;j++){
            if(arr[j+1]<arr[j]){
                temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
            }
        }
    }
}

File: Data-Structures-And-Algorithms/Sort/bubbleSort.go
package main

import (
	"fmt"
)

func main() {
	arr := []int{20, 13, 11, 48, 9, 25, 50, 1, 14, 100, 40, 25}
	ret := bubbleSort(arr)
	fmt.Println(ret)
}

/**
 * 冒泡法：两个for循环完成排序
 * 外循环：循环1次，可以把最大的数移动到最末尾，循环2次，可以把第二大的数放到末尾第二位置，
 *         因此有多少个数，就循环比较多少次（减1次）
 * 内循环：两个数比较后位置对调，然后在与后面的数比较对调
 *         对调多少次，比如最大的数100需要对调多少次移动到最后，总数减去当时外循环的位置
 */
func bubbleSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	var flag bool
	n := len(arr)
	for i := 0; i < n-1; i++ {
		for j := 0; j < n-1-i; j++ {
			if arr[j] > arr[j+1] {
				temp := arr[j]
				arr[j] = arr[j+1]
				arr[j+1] = temp
				flag = true
			}
		}
		if !flag {
			break
		} else {
			flag = false
		}
	}
	return arr
}
